File: app/src/main/java/com/petnbu/petnbu/db/CommentDao.java
package com.petnbu.petnbu.db;

import android.arch.lifecycle.LiveData;
import android.arch.persistence.room.Dao;
import android.arch.persistence.room.Insert;
import android.arch.persistence.room.OnConflictStrategy;
import android.arch.persistence.room.Query;
import android.arch.persistence.room.Transaction;
import android.arch.persistence.room.Update;
import android.support.annotation.Nullable;

import com.petnbu.petnbu.model.Comment;
import com.petnbu.petnbu.model.CommentEntity;
import com.petnbu.petnbu.model.CommentUI;

import java.util.List;

import timber.log.Timber;

@Dao
public abstract class CommentDao {

    @Transaction
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    public abstract void insert(CommentEntity... comments);

    @Transaction
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    public abstract void insert(List<CommentEntity> comments);

    public void insertFromComment(@Nullable Comment comment) {
        if (comment != null) {
            Timber.i("insertComment: %s", comment);
            insert(comment.toEntity());
        }
    }

    @Transaction
    public void insertListComment(List<Comment> comments){
        for (Comment comment : comments) {
            insertFromComment(comment);
        }
    }

    @Query("SELECT * FROM comments where id = :commentId")
    public abstract CommentEntity getCommentById(String commentId);

    @Query("SELECT comments.id, users.userId, users.avatar, users.name, comments.content, " +
            "comments.photo, comments.likeCount, comments.commentCount, comments.parentCommentId, " +
            "comments.parentFeedId, comments.localStatus, comments.timeCreated, " +
            "subComments.id as latestCommentId, subcomments.content as latestCommentContent, " +
            "subCommentUser.userId as latestCommentOwnerId, subCommentUser.name as latestCommentOwnerName, " +
            "subCommentUser.avatar as latestCommentOwnerAvatar " +
            "from comments " +
            "left join users on comments.ownerId = users.userId " +
            "left join comments as subComments on comments.latestCommentId = subComments.id " +
            "left join users as subCommentUser on subComments.ownerId = subCommentUser.userId " +
            "where comments.id in (:ids) " +
            "order by comments.timeCreated DESC")
    public abstract LiveData<List<CommentUI>> loadCommentsIncludeUploading(List<String> ids);

    @Update(onConflict = OnConflictStrategy.REPLACE)
    public abstract void update(CommentEntity commentEntity);

    @Query("UPDATE comments set id = :newId where id = :oldId")
    public abstract void updateCommentId(String oldId, String newId);
}

File: app/src/main/java/com/petnbu/petnbu/db/CommentDao.kt
package com.petnbu.petnbu.db

import android.arch.lifecycle.LiveData
import android.arch.persistence.room.*
import com.petnbu.petnbu.model.Comment
import com.petnbu.petnbu.model.CommentEntity
import com.petnbu.petnbu.model.CommentUI
import timber.log.Timber

@Dao
abstract class CommentDao {

    @Transaction
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    abstract fun insert(vararg comments: CommentEntity)

    @Insert(onConflict = OnConflictStrategy.IGNORE)
    abstract fun insertIfNotExists(vararg comments: CommentEntity)

    @Transaction
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    abstract fun insert(comments: List<CommentEntity>)

    fun insertFromComment(comment: Comment?) {
        Timber.i("insertComment: %s", comment)
        comment?.toEntity()?.let { insert(it) }
    }

    @Transaction
    open fun insertListComment(comments: List<Comment>) {
        comments.forEach({ insertFromComment(it) })
    }

    @Query("SELECT * FROM comments where id = :commentId")
    abstract fun getCommentById(commentId: String): CommentEntity?

    @Query("SELECT comments.id, users.userId, users.avatar, users.name, comments.content, " +
            "comments.photo, comments.likeCount, comments.isLiked, comments.likeInProgress, comments.commentCount, comments.parentCommentId, " +
            "comments.parentFeedId, comments.localStatus, comments.timeCreated, " +
            "subComments.id as latestCommentId, subcomments.content as latestCommentContent, " +
            "subCommentUser.userId as latestCommentOwnerId, subCommentUser.name as latestCommentOwnerName, " +
            "subCommentUser.avatar as latestCommentOwnerAvatar, subComments.photo as latestCommentPhoto " +
            "from comments " +
            "left join users on comments.ownerId = users.userId " +
            "left join comments as subComments on comments.latestCommentId = subComments.id " +
            "left join users as subCommentUser on subComments.ownerId = subCommentUser.userId " +
            "where comments.id in (:ids) or (comments.parentFeedId = :feedId and comments.localStatus = 1)" +
            "order by comments.timeCreated DESC")
    abstract fun loadFeedComments(ids: List<String>, feedId: String): LiveData<List<CommentUI>>

    //sub comments does not have latestComment fields
    @Query("SELECT comments.id, users.userId, users.avatar, users.name, comments.content, " +
            "comments.photo, comments.likeCount, comments.isLiked, comments.likeInProgress, comments.commentCount, comments.parentCommentId, " +
            "comments.parentFeedId, comments.localStatus, comments.timeCreated " +
            "from comments " +
            "left join users on comments.ownerId = users.userId " +
            "where comments.id in (:ids) or (comments.parentCommentId = :parentCommentId and comments.localStatus = 1)" +
            "order by comments.timeCreated DESC")
    abstract fun loadSubComments(ids: List<String>, parentCommentId: String): LiveData<List<CommentUI>>

    @Update(onConflict = OnConflictStrategy.REPLACE)
    abstract fun update(commentEntity: CommentEntity)

    @Query("UPDATE comments set id = :newId where id = :oldId")
    abstract fun updateCommentId(oldId: String, newId: String)
}
